compiler/rustc_monomorphize/src/partitioning/autodiff.rs

```diff
@@ -80,6 +101,20 @@ fn adjust_activity_to_abi<'tcx>(tcx: TyCtxt<'tcx>, fn_ty: Ty<'tcx>, da: &mut Vec
     }
 }
 
+fn count_scalar_fields<'tcx>(tcx: TyCtxt<'tcx>, ty: Ty<'tcx>) -> usize {
+    match ty.kind() {
+        ty::Float(_) | ty::Int(_) | ty::Uint(_) => 1,
+        ty::Adt(def, substs) if def.is_struct() => def
+            .non_enum_variant()
+            .fields
+            .iter()
+            .map(|f| count_scalar_fields(tcx, f.ty(tcx, substs)))
```

Comment:
Do you really want to recursively count and sum?

I think that anything behind a double indirection probably won't affect the size on the function abi, or?